packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,206 @@ import {
+	 * headings attribute.
+	 *
+	 * Since useSelect forces a re-render when its return value is shallowly
+	 * inequal to its prior call, we would be re-rendering this block every time
+	 * the stores change, even if the latest headings were deeply equal to the
+	 * ones saved in the block attributes.
+	 *
+	 * By returning null when they're equal, we reduce that to 2 renders: one
+	 * when there are new latest headings (and so it returns them), and one when
+	 * they haven't changed (so it returns null). As long as the latest heading
+	 * data remains the same, further calls of the useSelect callback will
+	 * continue to return null, thus preventing any forced re-renders.
+	 */
+	const latestHeadings = useSelect(
+		( select ) => {
 			const {
 				getBlockAttributes,
-				getBlockIndex,
 				getBlockName,
-				getBlockOrder,
+				getClientIdsWithDescendants,
+				__experimentalGetGlobalBlocksByName: getGlobalBlocksByName,
 			} = select( blockEditorStore );
 
-			const blockIndex = getBlockIndex( clientId );
-			const blockOrder = getBlockOrder();
-
-			// Calculate which page the block will appear in on the front-end by
-			// counting how many <!--nextpage--> tags precede it.
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// FIXME: @wordpress/block-library should not depend on @wordpress/editor.
+			// Blocks can be loaded into a *non-post* block editor, so to avoid
+			// declaring @wordpress/editor as a dependency, we must access its
+			// store by string. When the store is not available, editorSelectors
+			// will be null, and the block's saved markup will lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
+
+			const pageBreakClientIds = getGlobalBlocksByName( 'core/nextpage' );
+
+			const isPaginated = pageBreakClientIds.length !== 0;
+
+			// Get the client ids of all blocks in the editor.
+			const allBlockClientIds = getClientIdsWithDescendants();
+
+			// Calculate the page (of a paginated post) this block is part of.
+			let tocPage = 1;
+
+			if ( isPaginated ) {
+				// We can't use getBlockIndex because it only returns the index
+				// relative to sibling blocks.
+				const tocIndex = allBlockClientIds.indexOf( clientId );
+
+				for ( const [
+					blockIndex,
+					blockClientId,
+				] of allBlockClientIds.entries() ) {
+					// If we've reached blocks after the Table of Contents, we've
+					// finished calculating which page the block is on.
```

Comment:
This comment could do with more context. Why does the position of the table of contents matter?

Reply:
The Table of Contents block has an (off-by-default) toggle to only show headings from the current page (because posts can be paginated using Page Break blocks). Therefore, we need to know which page the Table of Contents is on so we know which headings to include.

Reply:
Your comment made me realize I could optimize the logic a bit by skipping this calculation when `onlyIncludeCurrentPage` is `false`, so I went ahead and added that optimization. I also tweaked a comment a few lines earlier in the code to better explain the purpose of the logic here.

Reply:
Thanks!